ramips:  Add support for Dlink DIR-878 R1 version

lior-pugatch opened the pull request:
This device has identical HW as DIR-878 A1, and has a flash map
As DIR-882 R1, which both are already supported.
Removed the DIR-882 R1 signature from the DIR-878 R1 as it did not help.
Method to upload the FW to 878 R1 will be updated in the device's wiki
page
Signed-off-by: lior-pugatch <buzz5800@gmail.com>

Commit:
ramips: Add support for DLink DIR-878 R1

Specifications:

    SoC: MT7621AT
    RAM: 128MB
    Flash: 16MB NOR SPI flash
    WiFi: MT7615N (2.4GHz) and MT7615N (5Ghz)
    LAN: 5x1000M
    Firmware layout is Uboot with extra 96 bytes in header
    Base PCB is AP-MTKH7-0002
    LEDs Power Green,Power Orange,Internet Green,Internet Orange
    LEDs "2.4G" Green & "5G" Green connected directly to wifi module
    Buttons Reset,WPS,WIFI

Flashing instructions:

Upload image via emergency recovery mode
Push and hold reset button (on the back of the device) until power led
starts flashing (about 10 secs or so) while powering the device on.
Give it ~30 seconds, to boot the recovery mode GUI
Connect your client computer to LAN1 of the device
Set your client IP address manually to 192.168.0.2 / 255.255.255.0.
Call the recovery page for the device at http://192.168.0.1
Use the provided emergency web GUI to upload and flash a new firmware to
the device. Some browsers/OS combinations are known not to work, so if
you don't see the percentage complete displayed and moving within a few
seconds, restart the procedure from scratch and try anoher one, ot try
the command line way.

This router has FW flavor R1 which has different Flash partitions than
the DIR-878 A1. i.e config2 partition is removed on R1

Signed-off-by: Lior Pugatch <buzz5800@gmail.com>

mans0n commented:
Hi, please remove unrelated changes from your patch.
And you should use your full name in the Signed-off-by tag.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#add_a_signed-off-by_with_your_real_name

Borromini commented:
@lior-pugatch You can copy the hardware details from the DIR-878 A1 commit message, also document in your commit message how to flash and ideally how to revert to stock as well.

It's good you'll update the wiki but the commit message [should contain flashing instructions too](https://openwrt.org/docs/guide-developer/device-support-policies#commit_message).

lior-pugatch commented:
Hi,

I've accidentally copied the target/linux/ramips/image/mk7621.mk from the main openwrt repo to the mirrored repo,
that's why there are so many changes there.

I've reverted this file and pushed it as another commit with just the relevan DIR-878 R1 addition.

is that good? 
or should I start a new PR to make things a bit cleaner?

I've updated the commit notes as offered here by @Borromini, leveraged the DIR-878 A1 commit notes
Also fixed the name issue.

Borromini commented:
Thanks, commit message is looking better. You should merge your commits locally, then force push your PR.

mans0n commented:
Please do not create merge commit. Squash the extra commits fixing problems in the first one into the original commit.
https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits

lior-pugatch commented:
Hi @mans0n .
Im a bit lost.
I've squashed the commits, and later forced pushed without doing a pull before the push to avoid a merging commit. 

It looks like my PR got closed.
Tried to follow the openwrt instructions on how to reopen a PR but I didnt undestand how I got to this situation

Please guide me further

mans0n commented:
Sorry, my github-fu is limited. I heard that some "wrong" push closes the PR automatically, but I can't find the reference right now.
Can you try following the instruction above?

xabolcs commented:
@lior-pugatch 

Check your `git reflog` output, searching for commit `0bdf8d1` and `b398f48`! Are there any commit between them?

GitHub knows b398f483b0df2df53ee23279e5b806b74d364e29 as the last good commit, but your reflog above could have more details.

- `git checkout add_support_dlink_dir_878_R1`
- `git reset --hard <last good commit from the reflog>`
- try squashing your commits again
- `git push <your github remote> add_support_dlink_dir_878_R1 --force` - check your GitHub remote with `git remote -v show` ... maybe `origin`?

lior-pugatch commented:
Trying to correctly squash commits

lior-pugatch commented:
Thx @xabolcs!

I was able to reopen the PR with the steps you have shared

mans0n commented:
@xabolcs Thanks for the detailed recovery procedure!

Thanks for the rebase, now I can have a look on this more closely.

mans0n commented on target/linux/ramips/dts/mt7621_dlink_dir-878-r1.dts:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "mt7621_dlink_dir-8xx-r1.dtsi"
```
Does this compile? It seems we don't have a file of this name?

lior-pugatch replied:
Before compiling locally I ran the command:
"git checkout v21.02.1"
taken from the instructions of building openwrt:
https://openwrt.org/docs/guide-developer/toolchain/use-buildsystem

It seems like branch 21.02.1 has these files. 
interesting why these files are not here on the mirror branch.
any thoughts?

Borromini replied:
@lior-pugatch Your PR is against master, right? You need to base it on master.

Master has [mt7621_dlink_dir-8xx.dtsi ](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_dlink_dir-8xx.dtsi) (note: no '-r1' suffix), no `mt7621_dlink_dir-8xx-r1.dtsi` though as mans0n says.

lior-pugatch replied:
When is it expected to have branch 21.02.1 merged to master?
is branch 21.02.1 considered "newer" than master?

xabolcs replied:
What is your interest in release 21.02?
Do you want see your DIR-878 R1 device supported by a 21.02 release?

Then do anything what the reviewers ask, and wait it for landing in `master`.
After that, pick that commit on `master` and _backport_ to the `openwrt-21.02` branch! With `git cherry-pick -x` and resolve the conflict (like this `.dtdi` mismatch).

For example see another backporting PR #4630:
* support for Xiaomi AIoT Router AC2350 landed first to `master` in commit 6c148116f778bfd2db88476cee4753b32538eafe from PR #4161, 
* and then was backported which was landed to `openwrt-21.02` branch as commit 06547e0a58daf768ff9776339cd31231dcd5c0ea.

Borromini replied:
@lior-pugatch Device support PRs are only accepted against master. As @xabolcs explained you can then request a backport to a stable release like 21.02. But only then; it needs to be in master first since that's where development happens.

mans0n replied:
Your work and test should be based on top of the master branch. Please do not just copy the patch based on the stable branch and say it's done.

mbc07 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -375,6 +375,15 @@ define Device/dlink_dir-878-a1
+  $(Device/dlink_dir-8xx-r1)
+  DEVICE_MODEL := DIR-878
+  DEVICE_VARIANT := R1
+  DEVICE_PACKAGES += kmod-usb3 kmod-usb-ledtrig-usbport
```
Is there any reason for including USB packages for the DIR-878 R1 considering the device doesn't have working USB ports (at least not without modding the hardware)?

hauke commented:
An other pull request adding support for Dlink DIR-878 R1 was merged in 42626aef676d2ca1ea1577b7a47ba8e4135f1a4c, see #9476
